kq_epic_post.md

```diff
@@ -0,0 +1,160 @@
+2. **Query identifiers change**. Each query is attached to a `fingerprint` as a unique identifier. The fingerprint's goal is to group similar queries that are structured the same way, but contain different IDs and field values. But sometimes a query's fingerprint changes. For example, in late November 2024, several of our affected queries appeared to have been solved, but at the same time new issues were being created with identical query structures and entry points but new fingerprints, [which we're tracking here](https://github.com/github/mysql-database-scorecard/issues/57). This meant that we had to be vigilant about cross-referencing old query issues to new ones.
+3. **Automations aren't perfect**. Killed query issues are regularly updated via automation with current performance data, which we used to determine impact and set priorities. We noticed multiple times that we stopped getting updates from the bot, which kept us from seeing if our data was accurate or current. Eventually the automation was fixed, but it was difficult to get this information in the meantime.
+
+### Plan your adventure
+With those challenges in mind, we began to strategize. We started by grouping our initial set of queries by possible solutions, including [batching](https://github.com/github/pull-requests/issues/14462) and [adding new indexes](https://github.com/github/pull-requests/issues/14461). However, there’s one bucket that stands out as the most fraught and the most impactful: our dependency on the `issues.state` field.
+
+A common trait among these killed queries was a non-performant join on the `issues` table. Because pull requests continue to track their state with an issue under the hood, querying a pull request by its state required us to also query the `issues` table. This resulted in queries that looked something like this:
+```
+SELECT
+    COUNT (*)
+  FROM
+    `pull_requests`
+    , `issues`
+  WHERE
+    `pull_requests`.`repository_id` = ?
+    AND (
+      `pull_requests`.`user_hidden` = ?
+      OR `pull_requests`.`user_id` = ?
+    )
+    AND `pull_requests`.`merged_at` IS NULL
+    AND `issues`.`state` = ?
+    AND `issues`.`pull_request_id` = `pull_requests`.`id`
+    AND `issues`.`repository_id` = `pull_requests`.`repository_id` LIMIT ?
+```
+The most basic question here is "Why can't a pull request have its own state?". The answer seemed like a clear win. Put a `state` column on `pull_requests` and remove the expensive join on issues `state`. Easy peasy.
+
+### Face the challengers
+Spoiler alert: It was not easy peasy. It was not even in the same realm as easy peasy. Like a labyrinth, the more turns we took, the more twists we discovered.
+
+#### Obstacle 1: When is a state not a state
```

Comment:
I really want a Lord of the Rings reference here since the other "Obstacle" titles have one, but I haven't thought of one yet.